ledger/participant-integration-api/src/test/suite/scala/platform/store/completions/PagedCompletionsReaderWithCacheSpec.scala

```diff
@@ -0,0 +1,328 @@
+import com.daml.ledger.participant.state.v1.Offset
+import com.daml.lf.data.Ref.Party
+import com.daml.logging.LoggingContext
+import com.daml.platform.store.completions.CompletionsDaoMock.{
+  GetAllCompletionsRequest,
+  GetFilteredCompletionsRequest,
+  Request,
+}
+import com.daml.platform.store.completions.RangeGenerator._
+import com.daml.platform.store.dao.CommandCompletionsTable
+import com.daml.platform.store.dao.CommandCompletionsTable.CompletionStreamResponseWithParties
+import org.scalatest.flatspec.AnyFlatSpec
+import org.scalatest.matchers.should.Matchers
+
+import scala.collection.immutable.SortedMap
+import scala.collection.mutable.ListBuffer
+import scala.concurrent.duration.Duration
+import scala.concurrent.{Await, ExecutionContext, Future}
+
+class PagedCompletionsReaderWithCacheSpec extends AnyFlatSpec with Matchers {
+
+  private implicit val executionContext: ExecutionContext =
+    ExecutionContext.fromExecutor(Executors.newFixedThreadPool(4))
+  private implicit val loggingContext: LoggingContext = LoggingContext.ForTesting
+
+  it should "correctly fetch historic data" in {
+
+    // GIVEN: completions reader and dao mock with 20 completions
+    val completionsDao = new CompletionsDaoMock(CompletionsDaoMock.genCompletions(20))
+    val reader = new TestablePagedCompletionsReaderWithCache(completionsDao, 10)
+
+    // WHEN: completions are requested for the first time
+    val cachedCompletions = await(
+      reader.getCompletionsPage(
+        genRange(9, 19),
+        ApplicationId.assertFromString("Application_1_Id"),
+        Set(Party.assertFromString("party1")),
+      )
+    )
+
+    // THEN returned completions size should be 10
+    cachedCompletions.size shouldBe 10
+    // AND all completions should be requested from database
+    completionsDao.runQueriesJournalToSeq() shouldBe Seq(
+      GetAllCompletionsRequest(genRange(9, 19))
+    )
+    // AND should be cached
+    reader.getCache shouldBe RangeCache(
+      genRange(9, 19),
+      10,
+      CompletionsDaoMock.genCompletionsMap(10, 20),
+    )
+
+    // WHEN: completions older than cached are fetched
+    val historicCompletions = await(
+      reader.getCompletionsPage(
+        Range(Offset.beforeBegin, genOffset(9)),
+        ApplicationId.assertFromString("Application_1_Id"),
+        Set(Party.assertFromString("party1")),
+      )
+    )
+
+    // THEN: 10 completions should be returned
+    historicCompletions.size shouldBe 10
+
+    // AND database should be queried with application and parties filter
+    completionsDao.runQueriesJournalToSeq() shouldBe Seq[Request](
+      GetAllCompletionsRequest(genRange(9, 19)),
+      GetFilteredCompletionsRequest(
+        Range(Offset.beforeBegin, genOffset(9)),
+        ApplicationId.assertFromString("Application_1_Id"),
+        Set(Party.assertFromString("party1")),
+      ),
+    )
+  }
+
+  it should "correctly fetch future (not cached yet) data and replace oldest cache with it" in {
```

Comment:
I think one corner case is not covered and actually also not implemented: what happens when cache contains elements 9..12 and a request comes in for values 15..19, I think we will end up with non-contiguous cache

Reply:
Currently we remove 9..12 and cache 15..19. It's limitation of current implementation.